internal/encryption/base.go

```diff
@@ -149,20 +140,16 @@ func (s *baseEncryption) decrypt(data []byte, validator func([]byte) error) ([]b
 			return nil, fmt.Errorf("unable to determine data structure during decryption: %w", verr)
 		}
 
-		methods, diags := s.buildTargetMethods(make(map[keyprovider.Addr][]byte))
-		if diags.HasErrors() {
-			// This cast to error here is safe as we know that at least one error exists
-			// This is also quite unlikely to happen as the constructor already has checked this code path
-			return nil, diags
-		}
 		// Yep, it's already decrypted
-		for _, method := range methods {
-			if method == nil {
-				// fallback allowed
+		for _, method := range s.encMethods {
+			if unencrypted.Is(method) {
+				if s.enforced {
+					return nil, fmt.Errorf("unable to use unencrypted method when enforced = true")
+				}
 				return data, nil
 			}
 		}
-		return data, fmt.Errorf("decrypted payload provided without fallback specified")
+		return nil, fmt.Errorf("decrypted payload provided without unencrypted method")
```

Comment:
I think we should improve this error message. As a user I wouldn't understand it (I almost didn't understand it now).

How about: "encountered unencrypted payload without unencrypted method configured"